### ISSUE TITLE
Assigning variables with %time within a function produces odd behavior

### ISSUE THREAD

**Issue by ternus** [ipython/ipython#9892]
```
import IPython; print(IPython.sys_info())
{'commit_hash': u'5c9c918',
 'commit_source': 'installation',
 'default_encoding': 'UTF-8',
 'ipython_path': '/usr/local/lib/python2.7/site-packages/IPython',
 'ipython_version': '5.1.0',
 'os_name': 'posix',
 'platform': 'Darwin-15.6.0-x86_64-i386-64bit',
 'sys_executable': '/usr/local/opt/python/bin/python2.7',
 'sys_platform': 'darwin',
 'sys_version': '2.7.12 (default, Jun 29 2016, 14:05:02) \n[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]'}
```

Assigning variables inside a magic function within a function scope doesn't work and exhibits very strange behavior around namespacing (see [this SO question](http://stackoverflow.com/questions/39067203/variables-defined-in-a-time-statement-inside-a-method-are-not-accessible-after). To reproduce:

``` .py
def func():
    a = 2
    get_ipython().magic("time b = 1")
    print(locals())
    print a
    print b
```

```
In [40]: func()
CPU times: user 3 µs, sys: 1 µs, total: 4 µs
Wall time: 5.96 µs
{'a': 2, 'b': 1}
2
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-40-08a2da4138f6> in <module>()
----> 1 func()

<ipython-input-39-9eb76d3bc6a2> in func()
      4     print(locals())
      5     print a
----> 6     print b
      7
      8

NameError: global name 'b' is not defined
```

**Comment by takluyver** [comment]
Sorry, that's a longstanding limitation in the way Python works. Local variables are not looked up from a dictionary the way global variables are, so dynamically executed code modifying the local namespace doesn't work as you'd expect.

**Comment by jayantj** [comment]
```
%time a = "abc"
print(a)
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 19.1 µs
abc
```
@takluyver This seems to work fine though. Is that because that would be a global variable lookup?

**Comment by takluyver** [comment]
Yup, precisely. It's only inside a function that it doesn't work.

**Comment by DanielGoldfarb** [comment]
I am seeing this same issue but NOT inside a function (unless I am missing something here) ...

    dino@DINO:~$ ipython
    Python 3.7.3 (default, Mar 27 2019, 22:11:17)
    Type 'copyright', 'credits' or 'license' for more information
    IPython 7.4.0 -- An enhanced Interactive Python. Type '?' for help.

    In [1]: import pandas as pd
    In [2]: pd.__version__
    Out[2]: '0.24.2'
    In [3]: df = pd.DataFrame({'A': [ 1, 2, 3],
       ...:                    'B': [ 4, 5, 6 ],
       ...:                    'C': [ 7, 8, 9 ] }, index = ['R1', 'R2', 'R3'])
    In [4]: df
    Out[4]:
        A  B  C
    R1  1  4  7
    R2  2  5  8
    R3  3  6  9

    In [5]: s = pd.Series([ 10, 20, 30], index=['A','B','C'])
    In [6]: s
    Out[6]:
    A    10
    B    20
    C    30
    dtype: int64

    In [7]: %timeit result = df - s
    1.28 ms ± 882 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

    In [8]: result
    ---------------------------------------------------------------------------
    NameError                                 Traceback (most recent call last)
    <ipython-input-8-0ac921c19f1a> in <module>
    ----> 1 result
    
    NameError: name 'result' is not defined

Note however, if I run `result = df -s` without %timeit, then it works fine:

    In [9]: result = df - s
    In [10]: result
    Out[10]:
        A   B   C
    R1 -9 -16 -23
    R2 -8 -15 -22
    R3 -7 -14 -21

And afterwards, I can run WITH %timeit and it also works fine.  But if I restart ipython, then as long as I only ever instantiate and assign the variable 'result' within %timeit, then I can repeatedly get the same error:

    In [8]: result
    ---------------------------------------------------------------------------
    NameError                                 Traceback (most recent call last)
    <ipython-input-8-0ac921c19f1a> in <module>
    ----> 1 result
    
    NameError: name 'result' is not defined

Am I missing something?  The variable result is global in this context isn't it?  It is not within a function.

**Comment by DanielGoldfarb** [comment]
For what it's worth, I tried the above `%timeit result = df -s` code in two different environments, and found the same error and behavior.  The environments are:

environment #1:

    In [14]: import IPython; print(IPython.sys_info())
    {'commit_hash': 'd774f565b',
     'commit_source': 'installation',
     'default_encoding': 'UTF-8',
     'ipython_path': '/home/dino/anaconda3/lib/python3.7/site-packages/IPython',
     'ipython_version': '7.4.0',
     'os_name': 'posix',
     'platform': 'Linux-4.4.0-17134-Microsoft-x86_64-with-debian-buster-sid',
     'sys_executable': '/home/dino/anaconda3/bin/python',
     'sys_platform': 'linux',
     'sys_version': '3.7.3 (default, Mar 27 2019, 22:11:17) \n[GCC 7.3.0]'}

environment #2:

    In [10]: import IPython; print(IPython.sys_info())
    {'commit_hash': '7f10f7bb3',
     'commit_source': 'installation',
     'default_encoding': 'UTF-8',
     'ipython_path': '/anaconda3/lib/python3.6/site-packages/IPython',
     'ipython_version': '6.4.0',
     'os_name': 'posix',
     'platform': 'Darwin-17.7.0-x86_64-i386-64bit',
     'sys_executable': '/anaconda3/bin/python',
     'sys_platform': 'darwin',
     'sys_version': '3.6.5 |Anaconda, Inc.| (default, Apr 26 2018, 08:42:37) \n'
                    '[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]'}

**Comment by DanielGoldfarb** [comment]
Hmm.  I just realized above you are talking about %time and I was talking about %timeit.
For what it's worth, I can reproduce the %timeit issue with a much simpler case; it appears that %timeit treats assignment as always to a variable that is local to %timeit itself.  So that variable is *never* seen outside of %timeit.  The cases I mentioned as working above, were only because I defined a different variable that was global, with the same name, and the same value.  Thus:

dino@DINO:~/code$ ipython
Python 3.7.3 (default, Mar 27 2019, 22:11:17)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.4.0 -- An enhanced Interactive Python. Type '?' for help.

    In [1]: b = 29
    In [2]: c = 32
    In [3]: %timeit a = b + c
    39.4 ns ± 0.585 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

    In [4]: a
    ---------------------------------------------------------------------------
    NameError                                 Traceback (most recent call last)
    <ipython-input-4-3f786850e387> in <module>
    ----> 1 a
    NameError: name 'a' is not defined

    In [5]: a = 20
    In [6]: %timeit a = b + c
    39.1 ns ± 0.0338 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

    In [7]: a
    Out[7]: 20    ## ! not 61 !
